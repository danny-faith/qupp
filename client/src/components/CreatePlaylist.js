import React, { Component } from 'react';
import { Row, Col, Button, Input } from 'react-materialize';
import classnames from 'classnames';
import { createPlaylist } from '../actions/authActions';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class CreatePlaylist extends Component {
    state = {
        name: '',
        errors: {}
    }
    onSubmit = (e) => {
        e.preventDefault();
        const newPlaylist = {
            name: this.state.name
        }
        this.props.createPlaylist(newPlaylist);
    }
    onChange = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }
    render() {
        const errors = this.state;

        return (
            <Row>
                <Col s={12}>
                    <h5>{this.props.title}</h5>
                    <form onSubmit={this.onSubmit}>
                        <Input
                            id={"name"}
                            className={classnames({
                                'invalid': errors.name
                            })} 
                            type="text"
                            name="name"
                            s={12}
                            label="Playlist name"
                            onChange={this.onChange}
                            value={this.state.name}
                        />
                        {/* {errors.name && (<p className="red-text col no-margin">{errors.playlist_name}</p>)} */}
                        <Button className="right">Create playlist</Button>
                    </form>
                </Col>
            </Row>
        )
    }
}

CreatePlaylist.propTypes = {
	createPlaylist: PropTypes.func.isRequired,
	auth: PropTypes.object.isRequired,
	errors: PropTypes.object.isRequired,
}

const mapStateToProps = (state) => ({
	auth: state.auth,
	errors: state.errors
});

export default connect(mapStateToProps, {createPlaylist })(CreatePlaylist);
