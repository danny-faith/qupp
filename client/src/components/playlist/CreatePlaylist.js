import React, { Component, Fragment } from 'react';
import { Button } from 'react-materialize';
import { createPlaylist } from '../../actions/playlistActions';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextFieldGroup from '../common/TextFieldGroup';

export class CreatePlaylist extends Component {
    state = {
        name: this.props.name || '',
        slug: this.props.slug || '',
        errors: {}
    }
    onSubmit = (e) => {
        e.preventDefault();
        const newPlaylist = {
            name: this.state.name,
            slug: this.state.slug
        }
        newPlaylist.id = (this.props.id) ? this.props.id : null;
        this.props.createPlaylist(newPlaylist);
    }
    onChange = (e) => {        
        const errors = this.state.errors;
		errors[e.target.name] = '';
		this.setState({
			[e.target.name]: e.target.value,
			errors
		});
	};
	componentWillReceiveProps = nextProps => {
		if (nextProps.errors) {
			this.setState({ errors: nextProps.errors });
		}
	};
	render() {
		const { errors } = this.state;

		return (
			<Fragment>
				<h5>{this.props.title}</h5>
				<form onSubmit={this.onSubmit}>
					<TextFieldGroup
						name="name"
						type="text"
						placeholder="Name"
						value={this.state.name}
						onChange={this.onChange}
						error={errors.name}
					/>
					<TextFieldGroup
						name="slug"
						type="text"
						placeholder="Playlist URL"
						value={this.state.slug}
						onChange={this.onChange}
						error={errors.slug}
						info="Please supply an easy to read URL for your playlist. No spaces, special characters or uppercase. Use underscores for spaces. E.g. daniels_party_jan"
					/>
					<Button className="right">{this.props.buttonText}</Button>
				</form>
			</Fragment>
		);
	}
}

CreatePlaylist.defaultProps = {
	buttonText: 'Create playlist'
};

CreatePlaylist.propTypes = {
	createPlaylist: PropTypes.func.isRequired,
	auth: PropTypes.object.isRequired,
	errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
	auth: state.auth,
	errors: state.errors
});

export default connect(
	mapStateToProps,
	{ createPlaylist }
)(CreatePlaylist);
