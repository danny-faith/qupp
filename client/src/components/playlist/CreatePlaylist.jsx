import React, { Fragment, useState, useMemo } from "react"
import { Button } from "react-materialize"
import { createPlaylist } from "../../actions/playlistActions"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import TextFieldGroup from "../common/TextFieldGroup"

function CreatePlaylist(props) {
	const [values, setValues] = useState({
		name: props.name || "",
		slug: props.slug || "",
	})
	const [errors, setErrors] = useState({})

	useMemo(() => {
		if (props.errors) {
			setErrors(props.errors)
		}
	}, [props.errors])

	const onSubmit = (e) => {
		e.preventDefault()
		const newPlaylist = {
			name: values.name,
			slug: values.slug,
		}
		newPlaylist.id = props.id ? props.id : null
		props.createPlaylist(newPlaylist)
	}

	const onChange = (e) => {
		setErrors({})
		setValues({ ...values, [e.target.name]: e.target.value })
	}

	return (
		<Fragment>
			<h5>{props.title}</h5>
			<form onSubmit={onSubmit}>
				<TextFieldGroup
					name="name"
					type="text"
					placeholder="Name"
					value={values.name}
					onChange={onChange}
					error={errors.name}
				/>
				<TextFieldGroup
					name="slug"
					type="text"
					placeholder="Playlist URL"
					value={values.slug}
					onChange={onChange}
					error={errors.slug}
					info="Please supply an easy to read URL for your playlist. No spaces, special characters or uppercase. Use underscores for spaces. E.g. daniels_party_jan"
				/>
				<Button className="right">{props.buttonText}</Button>
			</form>
		</Fragment>
	)
}

CreatePlaylist.defaultProps = {
	buttonText: "Create playlist",
}

CreatePlaylist.propTypes = {
	createPlaylist: PropTypes.func.isRequired,
	auth: PropTypes.object.isRequired,
	errors: PropTypes.object.isRequired,
}

const mapStateToProps = (state) => ({
	auth: state.auth,
	errors: state.errors,
})

export default connect(mapStateToProps, { createPlaylist })(CreatePlaylist)
