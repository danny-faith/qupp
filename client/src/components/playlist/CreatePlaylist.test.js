import React from 'react';
import { shallow, render, mount } from 'enzyme';
import { CreatePlaylist } from './CreatePlaylist';

describe('CreatePlaylist component', () => {
    let wrapper;
    const mockCreatePlaylistfn = jest.fn();
    const auth = {
        isAuthenticated: true,
    }
    beforeEach(() => {
        wrapper = shallow(
        <CreatePlaylist
            createPlaylist={mockCreatePlaylistfn}
            auth={auth}
            errors={{}}
        />)
    });

    it('should fully render(Full DOM) correctly', () => {
        const auth = {
            isAuthenticated: true,
        }
        const mockCreatePlaylistfn = jest.fn();
        const component = mount(
            <CreatePlaylist
                createPlaylist={mockCreatePlaylistfn}
                auth={auth}
                errors={{}}
            />
        );
        expect(component).toMatchSnapshot();
    });

    describe("CreatePlaylist inputs - ", () => {
        test("name input shows the expected text when user enters text", () => {
            const wrapper = mount(
                <CreatePlaylist
                createPlaylist={mockCreatePlaylistfn}
                auth={auth}
                errors={{}}
            />
            );
            wrapper.find('#name').at(1).simulate('change', 
                {target: 
                    {name: 'name', value: 'Daniels new playlist'}
                }
            );
            const nameInput = wrapper.find('#name').at(1).html();
            
            expect(nameInput).toEqual(expect.stringContaining('Daniels new playlist'));
        });
        
        test("Playlist URL input shows the expected text when user enters text", () => {
            const wrapper = mount(
                <CreatePlaylist
                createPlaylist={mockCreatePlaylistfn}
                auth={auth}
                errors={{}}
            />
            );
            wrapper.find('#name').at(1).simulate('change', 
                {target: 
                    {name: 'slug', value: 'sluggg'}
                }
            );
            const nameInput = wrapper.find('#slug').at(1).html();
            
            expect(nameInput).toEqual(expect.stringContaining('sluggg'));
        });
    });

})