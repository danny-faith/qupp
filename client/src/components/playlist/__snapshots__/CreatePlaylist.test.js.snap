// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePlaylist component should fully render(Full DOM) correctly 1`] = `
<CreatePlaylist
  auth={
    Object {
      "isAuthenticated": true,
    }
  }
  buttonText="Create playlist"
  createPlaylist={[MockFunction]}
  errors={Object {}}
>
  <Row>
    <div
      className="row"
    >
      <Col
        s={12}
      >
        <div
          className="col s12"
        >
          <h5 />
          <form
            onSubmit={[Function]}
          >
            <TextFieldGroup
              label="Name"
              name="name"
              onChange={[Function]}
              type="text"
              value=""
            >
              <Input
                checked={false}
                className="form-control form-control-lg"
                id="name"
                label="Name"
                name="name"
                onChange={[Function]}
                s={12}
                type="text"
                value=""
              >
                <div
                  className="col input-field s12"
                >
                  <input
                    checked={false}
                    className="form-control form-control-lg"
                    id="name"
                    name="name"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <label
                    className=""
                    htmlFor="name"
                  >
                    Name
                  </label>
                </div>
              </Input>
            </TextFieldGroup>
            <TextFieldGroup
              info="Please supply an easy to read URL for your playlist. No spaces, special characters or uppercase. Use underscores for spaces. E.g. daniels_party_jan"
              label="Slug"
              name="slug"
              onChange={[Function]}
              type="text"
              value=""
            >
              <Input
                checked={false}
                className="form-control form-control-lg"
                id="slug"
                label="Slug"
                name="slug"
                onChange={[Function]}
                s={12}
                type="text"
                value=""
              >
                <div
                  className="col input-field s12"
                >
                  <input
                    checked={false}
                    className="form-control form-control-lg"
                    id="slug"
                    name="slug"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <label
                    className=""
                    htmlFor="slug"
                  >
                    Slug
                  </label>
                </div>
              </Input>
              <p
                className="helper-text col s12"
              >
                Please supply an easy to read URL for your playlist. No spaces, special characters or uppercase. Use underscores for spaces. E.g. daniels_party_jan
              </p>
            </TextFieldGroup>
            <Button
              className="right"
              node="button"
            >
              <button
                className="btn right"
                disabled={false}
              >
                Create playlist
              </button>
            </Button>
          </form>
        </div>
      </Col>
    </div>
  </Row>
</CreatePlaylist>
`;
